### PR TITLE
Tests: Join pool for non-even ratio

### PR DESCRIPTION
**Pull request by mattverse**
Closes: #XXX

## What is the purpose of the change

Adds tests for join pool for non even ratio assets, so that the tests hit the side where even join is done to the maximal amount able to, and then do a single pool join for the remaining coins. 

All the results used within this test codes are hand derived, attached in comment the calculations done via wolfram

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: x/gamm/pool-models/balancer/amm_test.go [hunk 1]
```diff
@@ -799,6 +799,129 @@ func TestCalcJoinPoolShares(t *testing.T) {
...
+			),
+
+			expectShares: sdk.NewInt(2.5e12 + 1243750000000),
+			expectLiq: sdk.NewCoins(
+				sdk.NewInt64Coin("uosmo", 25_000),
+				sdk.NewInt64Coin("uatom", 50_000),
+			),
+		},
+		{
+			// join pool is first done to the extent where the ratio can be preserved, which is 25,000 uosmo and 12,500 uatom
```
Comment (stackman27) [inline x/gamm/pool-models/balancer/amm_test.go:885]: aah so the ratio is 2usmo for 1uatom?

File: x/gamm/pool-models/balancer/amm_test.go [hunk 2]
```diff
@@ -799,6 +799,129 @@ func TestCalcJoinPoolShares(t *testing.T) {
...
+			// then we perfrom single asset deposit for the remaining 37,500 uatom with the equation below
+			//
+			// For uatom:
+			// Expected output from Balancer paper (https://balancer.fi/whitepaper.pdf) using equation (25) on page 10:
+			// P_issued = P_supply * ((1 + (A_t * swapFeeRatio  / B_t))^W_t - 1)
+			// 609,374,990,000 = (100 * 10^18 + 1,250,000,000,000) *  (( 1 + (37,500 * (1 - (1 - 1/6) * 0.03) / 10,000,00,025,000))^1/6 - 1)
```
Comment (stackman27) [inline x/gamm/pool-models/balancer/amm_test.go:891]: do you happen to have the calculation for `1,250,000,000,000`?

File: x/gamm/pool-models/balancer/amm_test.go [hunk 3]
```diff
@@ -683,6 +683,129 @@ func TestCalcJoinPoolShares(t *testing.T) {
...
+				sdk.NewInt64Coin("uatom", 50_000),
+			),
+		},
+		{
+			// For uosmos and uatom
+			// join pool is first done to the extent where the ratio can be preserved, which is 25,000 uosmo and 25,000 uatom
+			// then we perfrom single asset deposit for the remaining 25,000 uatom with the equation below
+			// Expected output from Balancer paper (https://balancer.fi/whitepaper.pdf) using equation (25) on page 10:
+			// P_issued = P_supply * ((1 + (A_t * swapFeeRatio  / B_t))^W_t - 1)
+			// 1_243_750_000_000 = 100 * 10^18 *  (( 1 + (25000 * (1 - (1 - 0.5) * 0.01) / 1000000025000))^0.5 - 1)
```
[the remark this answers is not in the document]
Comment (mattverse) [reply x/gamm/pool-models/balancer/amm_test.go:849]: I think this is right?

[the remark this answers is not in the document]
Comment (mattverse) [reply x/gamm/pool-models/balancer/amm_test.go:849]: your wolfram uses 12,500 as A_t whilst we should be using 25,000

[the remark this answers is not in the document]
Comment (stackman27) [reply x/gamm/pool-models/balancer/amm_test.go:849]: aah gotchu!

File: x/gamm/pool-models/balancer/amm_test.go (hunk 1 above)
Comment (mattverse) [reply x/gamm/pool-models/balancer/amm_test.go:885]: we're not accounting weight here, the only thing that gets accounted here is token in and the amount of each pool asset

File: x/gamm/pool-models/balancer/amm_test.go (hunk 2 above)
Comment (mattverse) [reply x/gamm/pool-models/balancer/amm_test.go:892]: yeah! just added in the commetns

Comment (stackman27) [reply x/gamm/pool-models/balancer/amm_test.go:892]: Is there an equation for minimal shares that i'm not aware of?

Comment (mattverse) [reply x/gamm/pool-models/balancer/amm_test.go:867]: yeah check `MaximalExactRatioJoin`

**Review by ValarDragon — APPROVED** [review]
I checked the test vectors.

I'm overall pretty worried about error tolerances and #1733 , working on verifying more things in those.

Also test fvectors in this file are kind of crazy. Need to reduce boilerplate in them or move them to their own file

**Comment by mattverse** [comment]
current plan is to merge this and then create a new PR to remove all the noise in `amm_test.go`

**Pull request merged**